video/out/opengl/hwdec_surfacetexture.c

```diff
@@ -0,0 +1,343 @@
+        return -1;
+    }
+    p->surface = (*env)->NewGlobalRef(env, surface);
+    (*env)->DeleteLocalRef(env, surface);
+
+    p->hwctx = (struct mp_hwdec_ctx){
+        .driver_name = hw->driver->name,
+        .av_device_ref = create_mediacodec_device_ref(p->surface),
+    };
+    hwdec_devices_add(hw->devs, &p->hwctx);
+
+    return 0;
+}
+
+static void uninit(struct ra_hwdec *hw)
+{
+    struct priv_owner *p = hw->priv;
+    JNIEnv *env = get_jni_env();
+
+    JNI_CALL_VOID(p->texture, SurfaceTexture.release);
+    (*env)->DeleteGlobalRef(env, p->texture);
+
+    JNI_CALL_VOID(p->surface, Surface.release);
+    (*env)->DeleteGlobalRef(env, p->surface);
+
+    hwdec_devices_remove(hw->devs, &p->hwctx);
+    av_buffer_unref(&p->hwctx.av_device_ref);
+}
+
+static int mapper_init(struct ra_hwdec_mapper *mapper)
+{
+    struct priv *p = mapper->priv;
+    struct priv_owner *o = mapper->owner->priv;
+    GL *gl = ra_gl_get(mapper->ra);
+    JNIEnv *env = get_jni_env();
+
+    gl->GenTextures(1, &p->gl_texture);
+    gl->BindTexture(GL_TEXTURE_EXTERNAL_OES, p->gl_texture);
+    gl->TexParameteri(GL_TEXTURE_EXTERNAL_OES, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+    gl->TexParameteri(GL_TEXTURE_EXTERNAL_OES, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
+    gl->TexParameteri(GL_TEXTURE_EXTERNAL_OES, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+    gl->TexParameteri(GL_TEXTURE_EXTERNAL_OES, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+    gl->BindTexture(GL_TEXTURE_EXTERNAL_OES, 0);
+
+    JNI_CALL_VOID(o->texture, SurfaceTexture.attachToGLContext, p->gl_texture);
+    if (check_exception(mapper->owner, "SurfaceTexture.attachToGLContext"))
+        return -1;
+
+    mapper->dst_params = mapper->src_params;
+    mapper->dst_params.imgfmt = IMGFMT_RGB0;
+    mapper->dst_params.hw_subfmt = 0;
+
+    jfloatArray matrix = (*env)->NewFloatArray(env, 16);
+    p->matrix = (*env)->NewGlobalRef(env, matrix);
+    (*env)->DeleteLocalRef(env, matrix);
+
+    return 0;
+}
+
+static void mapper_uninit(struct ra_hwdec_mapper *mapper)
+{
+    struct priv *p = mapper->priv;
+    struct priv_owner *o = mapper->owner->priv;
+    GL *gl = ra_gl_get(mapper->ra);
+    JNIEnv *env = get_jni_env();
+
+    JNI_CALL_VOID(o->texture, SurfaceTexture.detachFromGLContext);
+    check_exception(mapper->owner, "SurfaceTexture.detachFromGLContext");
+
+    gl->DeleteTextures(1, &p->gl_texture);
+
+    (*env)->DeleteGlobalRef(env, p->matrix);
+}
+
+static void mapper_unmap(struct ra_hwdec_mapper *mapper)
+{
+    ra_tex_free(mapper->ra, &mapper->tex[0]);
+}
+
+static int mapper_map(struct ra_hwdec_mapper *mapper)
+{
+    struct priv *p = mapper->priv;
+    struct priv_owner *o = mapper->owner->priv;
+    JNIEnv *env = get_jni_env();
+
+    if (!p->gl_texture)
+        return -1;
+
+    AVMediaCodecBuffer *buffer = (AVMediaCodecBuffer *)mapper->src->planes[3];
+    av_mediacodec_release_buffer(buffer, 1);
+
+    JNI_CALL_VOID(o->texture, SurfaceTexture.updateTexImage);
```

Comment:
To map the mediacodec frame to a texture, I'm calling `MediaCodec#releaseOutputBuffer(render=1)` followed immediately by `SurfaceTexutre#updateTexImage`.

This seems to work fine on fast devices, but on slower ones I'm noticing video artifacts during playback. I think that must be because there's some latency before the texture becomes available, and we're supposed to use `SurfaceTexture#setOnFrameAvailableListener` before calling `updateTexImage`. That requires a java listener though, and happens asynchronously, so I'm not sure how best to integrate it into this JNI wrapper.

Reply:
> That requires a java listener though, and happens asynchronously, so I'm not sure how best to integrate it into this JNI wrapper.

Seems the only way to do this is to require the libmpv user to include a java file in their project. Here's an example:

- user would include [OnFrameAvailableListener.java](https://github.com/gopro/gopro-lib-node.gl/blob/master/libnodegl/android/java/OnFrameAvailableListener.java) in their java project
- libmpv would [fetch that class at runtime](https://github.com/gopro/gopro-lib-node.gl/blob/master/libnodegl/android/java/OnFrameAvailableListener.java) and [bind it to a C callback](https://github.com/gopro/gopro-lib-node.gl/blob/master/libnodegl/android_surface.c#L50)
- the C callback would [signal a condition](https://github.com/gopro/gopro-lib-node.gl/blob/master/libnodegl/android_surface.c#L456-L457) and the hwdec_surfacetexture mapper would [wait on the condition after calling releaseOutputBuffer](https://github.com/gopro/gopro-lib-node.gl/blob/master/libnodegl/android_surface.c#L394-L411)